File: Java/46. Permutations.java
// [1,2,3], with distinct numbers
// backtracking 
class Solution {
    public List<List<Integer>> permute(int[] nums) {        
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        helper(nums, new ArrayList<Integer>(), res, new HashSet<Integer>());
        return res;       
    }
    
    public void helper(int[] nums, List<Integer> curList, List<List<Integer>> res, HashSet<Integer> set){
        if(curList.size()==nums.length) res.add(new ArrayList<Integer>(curList));
        else{
            for(int i=0; i<nums.length; i++){
                if(!set.contains(nums[i])){
                    curList.add(nums[i]);
                    int lastIx = curList.size()-1;
                    set.add(nums[i]);
                    helper(nums, curList, res, set);
                    set.remove(nums[i]);
                    curList.remove(lastIx);
                }
            }
        }
    }
}

File: python/46. Permutations.py
class Solution:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ''' 
        # 1) Interation
        result = [nums]
        for i in range(len(nums)-1):
            for one in result[:]:
                for j in range(i+1, len(nums)):
                    result.append(one[:i] + one[j:] + one[i:j])
        return result      
        '''
        
        #  2) backtracking
        res = []        
        def helper(nums, temp):
            if len(temp) == len(nums):
                res.append(temp[:])
                
            for i in range(len(nums)):
                if nums[i] in temp:
                    continue
                temp.append(nums[i])
                helper(nums, temp)
                temp.pop()
                
        helper(nums,[])
        return res
    
    
    # 3) dynamic programming
        def insert_num(temp, num):
            if temp == []:
                temp.append([num])
                return temp

            res = []
            for i in temp:
                for j in range(len(i) + 1):
                    i_copy = list(i)
                    i_copy.insert(j, num)
                    res.append(i_copy)
            return res

        temp = []
        for i in nums:
            temp = insert_num(temp, i)
            
        return temp
